<?xml version="1.0"?>

<launch>

  <!-- Obstacle parameters  -->
  <arg name="n_of_obstacles" default="3" />
  <arg name="obstacleNames" default="[bebop0,bebop2,bebop3]" />
  <arg name="secs_per_rev" default="17" />
  <arg name="ra" default="0.3" />
  <arg name="rbar" default="0.5" />
  <arg name="rd" default="0.8" />

  <!-- ADP controller parameters  --> 
  <arg name="init_desPose" default="[1.0 1.0 1.0]" />
  
  
  <!-- Low level velocity tracker parameters  --> 
  <arg name="linVelGains" default="[0.5, 0.12, 0.2]" />
  <arg name="angVelGains" default="[2, 0.7, 0.8]" />
  <arg name="errorDerivativeBufferSize" default="3" />
  <arg name="errorIntegralBufferSize" default="0.5" />

  <!-- Nodes for obstacles  --> 	
  <node name="obstacle1" pkg="adp_staf_pp_dynamic" type="obstacle_node">
    <param name="bebopName" value="bebop0" />
    <param name="loopRate" value="300" />
    <param name="primary_centerx" value="0" />
    <param name="primary_centery" value="0" />
    <param name="primary_radius1" value="1.5" />
    <param name="primary_radius2" value="1.5" />
    <param name="primary_height" value="1" />
    <param name="secs_per_rev" value="$(arg secs_per_rev)" />
    <param name="phase_shift" value="0" />
  </node>
  
  <node name="obstacle2" pkg="adp_staf_pp_dynamic" type="obstacle_node">
    <param name="bebopName" value="bebop2" />
    <param name="loopRate" value="300" />
    <param name="primary_centerx" value="0.5" />
    <param name="primary_centery" value="0" />
    <param name="primary_radius1" value="1" />
    <param name="primary_radius2" value="0.5" />
    <param name="primary_height" value="1" />
    <param name="secs_per_rev" value="$(arg secs_per_rev)" />
    <param name="phase_shift" value="3.14" />
    <param name="reverse" value="true" />
  </node>

  <node name="obstacle3" pkg="adp_staf_pp_dynamic" type="obstacle_node">
    <param name="bebopName" value="bebop3" />
    <param name="loopRate" value="300" />
    <param name="primary_centerx" value="-1.5" />
    <param name="primary_centery" value="0" />
    <param name="primary_radius1" value="1" />
    <param name="primary_radius2" value="0.5" />
    <param name="primary_height" value="1" />
    <param name="secs_per_rev" value="$(arg secs_per_rev)" />
    <param name="phase_shift" value="3.14" />
  </node>
  
  <!-- Node for controller  --> 
  <node name="agent_controller" pkg="adp_staf_pp_dynamic" type="agent_nonauto_controller_node" output="screen">
    <param name="bebopName" value="bebop4" />
    <param name="n_of_obstacles" value="$(arg n_of_obstacles)" /> 
    <param name="ra" value="$(arg ra)" /> 
    <param name="rbar" value="$(arg rbar)" /> 
    <param name="rd" value="$(arg rd)" /> 
    <rosparam param="init_desPose" subst_value="True">$(arg init_desPose)</rosparam>
  </node>
  
  <!-- Node for simulating agent (for checking controller performance before actually flying)  --> 
  <node name="agent_sim" pkg="adp_staf_pp_dynamic" type="agent_sim_node" output="screen">
    <param name="bebopName" value="bebop4" />
  </node>

  <!-- Node for low level velocity tracking controller  --> 
  <node name="agent_low_level_control" pkg="adp_staf_pp_dynamic" type="bebop_low_level_controller_node" output="screen">
    <rosparam param="linVelGains" subst_value="True">$(arg linVelGains)</rosparam><!-- gain for linear velocity error tracking (Kp,Kd,Ki)-->
    <rosparam param="angVelGains" subst_value="True">$(arg angVelGains)</rosparam><!-- gain for angular velocity error tracking (Kp,Kd,Ki)-->
    <param name="bebopName" value="bebop4" />
  </node>
  <node name="obstacle1_low_level_control" pkg="adp_staf_pp_dynamic" type="bebop_low_level_controller_node" output="screen">
    <rosparam param="linVelGains" subst_value="True">$(arg linVelGains)</rosparam><!-- gain for linear velocity error tracking (Kp,Kd,Ki)-->
    <rosparam param="angVelGains" subst_value="True">$(arg angVelGains)</rosparam><!-- gain for angular velocity error tracking (Kp,Kd,Ki)-->
    <param name="bebopName" value="bebop1" />
  </node>  
  <node name="obstacle2_low_level_control" pkg="adp_staf_pp_dynamic" type="bebop_low_level_controller_node" output="screen">
    <rosparam param="linVelGains" subst_value="True">$(arg linVelGains)</rosparam><!-- gain for linear velocity error tracking (Kp,Kd,Ki)-->
    <rosparam param="angVelGains" subst_value="True">$(arg angVelGains)</rosparam><!-- gain for angular velocity error tracking (Kp,Kd,Ki)-->
    <param name="bebopName" value="bebop2" />
  </node>
  <node name="obstacle3_low_level_control" pkg="adp_staf_pp_dynamic" type="bebop_low_level_controller_node" output="screen">
    <rosparam param="linVelGains" subst_value="True">$(arg linVelGains)</rosparam><!-- gain for linear velocity error tracking (Kp,Kd,Ki)-->
    <rosparam param="angVelGains" subst_value="True">$(arg angVelGains)</rosparam><!-- gain for angular velocity error tracking (Kp,Kd,Ki)-->
    <param name="bebopName" value="bebop3" />
  </node> 
   
  
</launch>

